packages/server/modules/workspaces/events/eventListener.ts

```diff
@@ -254,44 +265,104 @@ export const onWorkspaceRoleDeletedFactory =
     await deleteWorkspaceSeat({ userId, workspaceId })
   }
 
+export const onWorkspaceSeatUpdatedFactory =
+  (deps: {
+    getWorkspaceSeatTypeToProjectRoleMapping: GetWorkspaceSeatTypeToProjectRoleMapping
+    queryAllWorkspaceProjects: QueryAllWorkspaceProjects
+    setStreamCollaborator: SetStreamCollaborator
+    getWorkspaceWithPlan: GetWorkspaceWithPlan
+    getWorkspaceRoleForUser: GetWorkspaceRoleForUser
+  }) =>
+  async (params: EventPayload<typeof WorkspaceEvents.SeatUpdated>) => {
+    const { seat, updatedByUserId } = params.payload
+    const { userId, type: seatType, workspaceId } = seat
+
+    const [workspace, role] = await Promise.all([
+      deps.getWorkspaceWithPlan({ workspaceId }),
+      deps.getWorkspaceRoleForUser({ userId, workspaceId })
+    ])
+    if (!workspace || !role) return
+
+    // Only new plans only rely on seat types
+    const isNewPlan = workspace.plan && isNewPaidPlanType(workspace.plan.name)
+    if (!isNewPlan) {
+      return
+    }
+
+    const { allowed: allowedProjectRoles, default: defaultProjectRoles } =
+      await deps.getWorkspaceSeatTypeToProjectRoleMapping({
+        workspaceId
+      })
+
+    // Ensure project roles are valid on seat type switch
+    for await (const projectsPage of deps.queryAllWorkspaceProjects({
+      workspaceId,
+      userId
+    })) {
+      await Promise.all(
+        projectsPage.map(async ({ id: projectId, role: originalProjectRole }) => {
+          const disallowedProjectRole =
+            originalProjectRole &&
+            !allowedProjectRoles[seatType].includes(originalProjectRole)
+          if (!disallowedProjectRole) return
+
+          const newRole = defaultProjectRoles[seatType]
+          await deps.setStreamCollaborator(
+            {
+              streamId: projectId,
+              userId,
+              role: newRole,
+              setByUserId: updatedByUserId
+            },
+            { trackProjectUpdate: false, skipAuthorization: true }
+          )
+        })
+      )
+    }
+  }
+
 export const onWorkspaceRoleUpdatedFactory =
-  ({
-    getWorkspaceRolesAllowedProjectRoles,
-    queryAllWorkspaceProjects,
-    setStreamCollaborator,
-    getStreamsCollaboratorCounts
-  }: {
-    getWorkspaceRolesAllowedProjectRoles: GetWorkspaceRolesAllowedProjectRolesFactory
+  (deps: {
```

Comment:
not a blocker for this PR, just a comment, having these kind of changes in a code review is annoying and makes no differences in the functionality